File: LeetCode/1559. Detect Cycles in 2D Grid/solution.py
class Solution:
    def containsCycle(self, grid: List[List[str]]) -> bool:
        n, m = len(grid), len(grid[0])
        tot = n*m 
        parent = list(range(tot))
        def find(u):
            if parent[u] != u:
                parent[u] = find(parent[u])
            return parent[u]
        def connect(u,v):
            ru, rv = find(u), find(v)
            if ru == rv:
                return True
            parent[ru] = rv 
            return False

        for i in range(n):
            for j in range(m):
                if i and grid[i][j] == grid[i-1][j]:
                    if connect(i*m+j,(i-1)*m+j):
                        return True
                if j and grid[i][j] == grid[i][j-1]:
                    if connect(i*m+j,i*m+j-1):
                        return True
        return False


import sys
sys.setrecursionlimit(10**6)

class Solution:
    def containsCycle(self, grid: List[List[str]]) -> bool:
        n, m = len(grid), len(grid[0])
        seen = [[False]*m for _ in range(n)]

        def dfs(i,j,parent):
            
            seen[i][j] = True
            for ni, nj in [(0,1),(0,-1),(1,0),(-1,0)]:
                ni += i 
                nj += j
                if ni < 0 or ni >= n or nj < 0 or nj>= m or (ni,nj) == parent or grid[ni][nj] != grid[i][j]:
                    continue
                if seen[ni][nj] or dfs(ni,nj,(i,j)):
                    return True
            return False


        for r in range(n):
            for c in range(m):
                if not seen[r][c]:
                    if dfs(r,c,(-1,-1)):
                        return True
                    # print(r,c,seen,depth)
        return False

File: LeetCode/1559. Detect Cycles in 2D Grid/solution.cpp
using pp = pair<int,int>;

vector<pp> dxy = {{-1,0},{1,0},{0,-1},{0,1}};

class Solution {
public:
    int n, m;
    vector<vector<bool>> visited;
    vector<vector<char>> grid;
    bool tarjan(int i, int j, int pi, int pj){
        visited[i][j] = true;
        for (auto [dx,dy] : dxy){
            dx += i; dy += j;
            if (dx == pi && dy == pj)
                continue;
            if (dx < 0 || dx >= n || dy < 0 || dy >= m)
                continue;
            if (grid[i][j] != grid[dx][dy])
                continue;
            if (visited[dx][dy]){
                // cout << i << " " << j << " " << pi << " " << pj << " " << dx << " " << dy << endl;
                return true;
            }
                
            if (tarjan(dx,dy,i,j))
                return true;
        }
        return false;
    }
    bool containsCycle(vector<vector<char>>& grid) {
        this->grid = grid;
        n = grid.size(), m = grid[0].size();
        visited.resize(n);
        for (int i = 0; i < n; i++)
            visited[i].assign(m,false);
        for (int i = 0; i < n; i++){
            for (int j = 0; j < m; j++){
                if (!visited[i][j] && tarjan(i,j,-1,-1))
                    return true;
            }
        }
        return false;
    }
};